modules/nextflow/src/main/groovy/nextflow/extension/op/ContextRunPerThread.groovy

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2013-2024, Seqera Labs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package nextflow.extension.op
+
+
+import groovy.transform.CompileStatic
+import groovy.util.logging.Slf4j
+import nextflow.prov.OperatorRun
+/**
+ * Implements an operator context that binds a new run to the current thread
```

Comment:
Worth commenting here that a "thread" here refers to a DataflowProcessor (i.e. input channel).

My custom tests on `join` and `mix` are working correctly, so I'm assuming that each DataflowProcessor uses the same thread for all of its runs (as long as maxForks is 1).

For operators like `collate` and `groupTuple` that use only one DP, the run-per-thread context essentially allows you to manually override how the provenance links are recorded. They are also working correctly with my tests.

Reply:
Even better would be to replace the thread-local with a `Map<DataflowProcessor,OperatorLink>`, but not a big deal if that would be too complicated